File: vciplreportsv1001/vcipl1001/report/pending_sales_order_report/pending_sales_order_report.py
import copy
import frappe
from frappe import _
from frappe.utils import flt, date_diff


def execute(filters=None):

    if not filters:
        return [], [], None, []

    validate_filters(filters)

    columns = get_columns(filters)
    conditions = get_conditions(filters)
    data = get_data(conditions, filters)

    # ✅ keep only pending rows
    data = [d for d in data if flt(d.pending_qty) > 0]

    if not data:
        return columns, [], None, []

    data, chart_data, totals = prepare_data(data, filters)

    report_summary = [
        {"label": _("Ordered Qty"), "value": totals["qty"], "datatype": "Float"},
        {"label": _("Delivered Qty"), "value": totals["delivered_qty"], "datatype": "Float"},
        {"label": _("Qty to Deliver"), "value": totals["pending_qty"], "datatype": "Float"},
        {"label": _("Order Amount"), "value": totals["amount"], "datatype": "Currency"},
        {"label": _("Delivered Amount"), "value": totals["delivered_amount"], "datatype": "Currency"},
        {"label": _("Pending Amount"), "value": totals["pending_amount"], "datatype": "Currency"},
    ]

    return columns, data, None, chart_data, report_summary


# ---------------- VALIDATION ----------------

def validate_filters(filters):

    if filters.get("from_date") and filters.get("to_date"):
        if date_diff(filters.to_date, filters.from_date) < 0:
            frappe.throw(_("To Date cannot be before From Date."))


# ---------------- CONDITIONS ----------------

def get_conditions(filters):

    conditions = ""

    if filters.get("from_date") and filters.get("to_date"):
        conditions += " AND so.transaction_date BETWEEN %(from_date)s AND %(to_date)s"

    if filters.get("company"):
        conditions += " AND so.company = %(company)s"

    if filters.get("sales_order"):
        conditions += " AND so.name IN %(sales_order)s"

    if filters.get("status"):
        conditions += " AND so.status IN %(status)s"

    if filters.get("warehouse"):
        conditions += " AND soi.warehouse = %(warehouse)s"

    return conditions


# ---------------- DATA ----------------

def get_data(conditions, filters):

    data = frappe.db.sql(f"""
        SELECT
            so.transaction_date AS date,
            so.name AS sales_order,
            so.status,
            c.customer_name AS customer,
            soi.item_code,
            soi.item_name,
            soi.qty,
            soi.delivered_qty,
            (soi.qty - soi.delivered_qty) AS pending_qty,
            soi.delivery_date,
            soi.warehouse,
            soi.base_amount AS amount,
            (soi.delivered_qty * soi.base_rate) AS delivered_amount,
            (soi.base_amount - (soi.delivered_qty * soi.base_rate)) AS pending_amount
        FROM `tabSales Order` so
        INNER JOIN `tabSales Order Item` soi ON soi.parent = so.name
        INNER JOIN `tabCustomer` c ON c.name = so.customer
        WHERE so.docstatus = 1
        AND so.status NOT IN ('Completed','Closed','Stopped','On Hold')
        {conditions}
    """, filters, as_dict=True)

    # stock lookup
    bins = frappe.db.sql("""
        SELECT item_code, warehouse, actual_qty
        FROM `tabBin`
    """, as_dict=True)

    bin_map = {(b.item_code, b.warehouse): b.actual_qty for b in bins}

    so_popup = {}

    for row in data:

        available = bin_map.get((row.item_code, row.warehouse), 0)

        ratio = (row.pending_qty / row.qty * 100) if row.qty > 0 else 0
        row["fill_ratio"] = round(ratio, 2)

        so_popup.setdefault(row.sales_order, []).append({
            "item_name": row.item_name,
            "pending_qty": row.pending_qty,
            "available_qty": available,
            "ratio": round(ratio, 2),
            "delivery_date": str(row.delivery_date) if row.delivery_date else ""
        })

    for row in data:
        row["pending_popup"] = frappe.as_json(so_popup.get(row.sales_order, []))

    return data


# ---------------- PREPARE ----------------

def prepare_data(data, filters):

    sales_order_map = {}

    totals = {
        "qty": 0,
        "delivered_qty": 0,
        "pending_qty": 0,
        "amount": 0,
        "delivered_amount": 0,
        "pending_amount": 0,
    }

    for row in data:

        totals["qty"] += flt(row.qty)
        totals["delivered_qty"] += flt(row.delivered_qty)
        totals["pending_qty"] += flt(row.pending_qty)
        totals["amount"] += flt(row.amount)
        totals["delivered_amount"] += flt(row.delivered_amount)
        totals["pending_amount"] += flt(row.pending_amount)

        if filters.get("group_by_so"):

            so = row.sales_order

            if so not in sales_order_map:
                sales_order_map[so] = copy.deepcopy(row)
            else:
                sales_order_map[so].qty += flt(row.qty)
                sales_order_map[so].delivered_qty += flt(row.delivered_qty)
                sales_order_map[so].pending_qty += flt(row.pending_qty)
                sales_order_map[so].amount += flt(row.amount)
                sales_order_map[so].delivered_amount += flt(row.delivered_amount)
                sales_order_map[so].pending_amount += flt(row.pending_amount)

    chart_data = {
        "data": {
            "labels": [_("Pending Amount"), _("Delivered Amount")],
            "datasets": [{"values": [totals["pending_amount"], totals["delivered_amount"]]}],
        },
        "type": "donut",
        "height": 300,
    }

    if filters.get("group_by_so"):
        return list(sales_order_map.values()), chart_data, totals

    return data, chart_data, totals


# ---------------- COLUMNS ----------------

def get_columns(filters):

    qty_label = _("Ordered Qty") if filters.get("group_by_so") else _("Qty")

    return [
        {"label": _("Date"), "fieldname": "date", "fieldtype": "Date"},
        {"label": _("Sales Order"), "fieldname": "sales_order", "fieldtype": "Link", "options": "Sales Order"},
        {"label": _("Pending Delivery"), "fieldname": "pending_delivery"},
        {"label": _("Status"), "fieldname": "status"},
        {"label": _("Customer"), "fieldname": "customer"},

        {"label": qty_label, "fieldname": "qty", "fieldtype": "Float"},
        {"label": _("Delivered Qty"), "fieldname": "delivered_qty", "fieldtype": "Float"},
        {"label": _("Qty to Deliver"), "fieldname": "pending_qty", "fieldtype": "Float"},
        {"label": _("Fill Ratio %"), "fieldname": "fill_ratio", "fieldtype": "Float"},

        {"label": _("Order Amount"), "fieldname": "amount", "fieldtype": "Currency"},
        {"label": _("Delivered Amount"), "fieldname": "delivered_amount", "fieldtype": "Currency"},
        {"label": _("Pending Amount"), "fieldname": "pending_amount", "fieldtype": "Currency"},

        {"fieldname": "pending_popup", "hidden": 1},
    ]

File: vciplreportsv1001/vcipl1001/report/pending_sales_order_report/pending_sales_order_report.js
// frappe.query_reports["Pending Sales Order Report"] = {

//     filters: [
//         {
//             fieldname: "company",
//             label: __("Company"),
//             fieldtype: "Link",
//             options: "Company",
//             default: frappe.defaults.get_user_default("Company"),
//             reqd: 1
//         },

//         {
//             fieldname: "from_date",
//             label: __("From Date"),
//             fieldtype: "Date",
//             default: frappe.datetime.year_start()
//         },

//         {
//             fieldname: "to_date",
//             label: __("To Date"),
//             fieldtype: "Date",
//             default: frappe.datetime.year_end()
//         },

//         {
//             fieldname: "sales_order",
//             label: __("Sales Order"),
//             fieldtype: "MultiSelectList",
//             options: "Sales Order",
//             get_data(txt) {
//                 return frappe.db.get_link_options("Sales Order", txt);
//             }
//         },

//         { fieldname: "customer", label: __("Customer"), fieldtype: "Link", options: "Customer" },

//         {
//             fieldname: "status",
//             label: __("Status"),
//             fieldtype: "MultiSelectList",
//             default: ["To Deliver and Bill"]
//         },

//         { fieldname: "warehouse", label: __("Warehouse"), fieldtype: "Link", options: "Warehouse" },

//         { fieldname: "group_by_so", label: __("Group by Sales Order"), fieldtype: "Check", default: 1 }
//     ],

//     formatter(value, row, column, data, default_formatter) {

//         value = default_formatter(value, row, column, data);

//         if (column.fieldname === "pending_qty" && data?.pending_qty > 0) {
//             value = `<span style="color:red;font-weight:600">${value}</span>`;
//         }

//         if (column.fieldname === "pending_delivery") {
//             return `<a style="font-weight:bold;color:#1674E0;cursor:pointer"
//                 onclick='frappe.query_reports["Pending Sales Order Report"]
//                 .show_popup(${data.pending_popup}, "${data.customer}", "${data.sales_order}")'>
//                 View Pending
//             </a>`;
//         }

//         return value;
//     },

//     show_popup(rows, customer, sales_order) {

//         let html = `
//         <div id="pending-popup">

//         <h4>Customer: ${customer}</h4>
//         <h5>Sales Order: ${sales_order}</h5>

//         <div id="print-container">
//         <table class="table table-bordered">
//         <tr>
//             <th>Item Code</th>   <!-- ✅ ADDED -->
//             <th>Item Name</th>
//             <th>Pending</th>
//             <th>Available</th>
//             <th>Date</th>
//         </tr>`;

//         rows.forEach(r => {

//             if (r.pending_qty <= 0) return;

//             html += `
//             <tr>
//                 <td>${r.item_code}</td> <!-- ✅ ADDED -->
//                 <td>${r.item_name}</td>
//                 <td style="color:red;font-weight:bold">${r.pending_qty}</td>
//                 <td style="color:green;font-weight:bold">${r.available_qty}</td>
//                 <td>${r.delivery_date}</td>
//             </tr>`;
//         });

//         html += `
//         </table>
//         </div>

//         <button class="btn btn-primary" onclick="print_pending_popup()">Print</button>
//         </div>`;

//         frappe.msgprint({
//             title: "Pending Items",
//             message: html,
//             wide: true
//         });
//     }
// };


// window.print_pending_popup = function () {

//     let popup = document.getElementById("pending-popup").cloneNode(true);

//     let btn = popup.querySelector("button");
//     if (btn) btn.remove();

//     let content = popup.outerHTML;

//     let w = window.open("", "_blank");

//     w.document.open();

//     w.document.write(`
//         <!DOCTYPE html>
//         <html>
//         <head>
//             <title>Pending Items</title>

//             <style>

//                 @page {
//                     size: A4;
//                     margin: 10mm;
//                 }

//                 body{
//                     font-family: Arial, sans-serif;
//                     margin:0;
//                     width:100%;
//                 }

//                 h4{
//                     margin-bottom:5px;
//                 }

//                 h5{
//                     margin-top:0;
//                     margin-bottom:15px;
//                 }

//                 table{
//                     width:100%;
//                     border-collapse:collapse;
//                     table-layout:auto;
//                 }

//                 th, td{
//                     border:1px solid black;
//                     padding:8px;
//                     text-align:left;
//                     word-wrap:break-word;
//                 }

//                 th{
//                     background:#f2f2f2;
//                 }

//             @media print{

//                 body{
//                     width:100%;
//                 }

//                 #print-container{
//                     max-height: none !important;
//                     overflow: visible !important;
//                 }

//                 table{
//                     width:100%;
//                     border-collapse: collapse;
//                 }

//                 tr{
//                     page-break-inside: avoid;
//                 }

//                 thead{
//                     display: table-header-group;
//                 }

//             }

//             </style>

//         </head>

//         <body>

//         ${content}

//         </body>
//         </html>
//     `);

//     w.document.close();

//     w.onload = function () {
//         w.focus();
//         w.print();
//     };
// };


frappe.query_reports["Pending Sales Order Report"] = {

    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("Company"),
            reqd: 1
        },

        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.year_start()
        },

        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.year_end()
        },

        {
            fieldname: "sales_order",
            label: __("Sales Order"),
            fieldtype: "MultiSelectList",
            options: "Sales Order",
            get_data(txt) {
                return frappe.db.get_link_options("Sales Order", txt);
            }
        },

        { fieldname: "customer", label: __("Customer"), fieldtype: "Link", options: "Customer" },

        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "MultiSelectList",
            default: ["To Deliver and Bill"]
        },

        { fieldname: "warehouse", label: __("Warehouse"), fieldtype: "Link", options: "Warehouse" },

        { fieldname: "group_by_so", label: __("Group by Sales Order"), fieldtype: "Check", default: 1 }
    ],

    // Cache row data in memory, keyed by a safe generated string.
    // We NEVER embed raw customer/item text (which may contain apostrophes,
    // quotes, etc. e.g. "Chef's Lagan") into an onclick attribute anymore.
    _pending_cache: {},
    _pending_key_counter: 0,

    formatter(value, row, column, data, default_formatter) {

        value = default_formatter(value, row, column, data);

        if (column.fieldname === "pending_qty" && data?.pending_qty > 0) {
            value = `<span style="color:red;font-weight:600">${value}</span>`;
        }

        if (column.fieldname === "pending_delivery") {

            let report = frappe.query_reports["Pending Sales Order Report"];
            report._pending_key_counter += 1;
            let key = "row_" + report._pending_key_counter;

            report._pending_cache[key] = {
                rows: data.pending_popup,
                customer: data.customer,
                sales_order: data.sales_order
            };

            return `<a style="font-weight:bold;color:#1674E0;cursor:pointer"
                onclick='frappe.query_reports["Pending Sales Order Report"].show_popup_by_key("${key}")'>
                View Pending
            </a>`;
        }

        return value;
    },

    show_popup_by_key(key) {
        let cached = frappe.query_reports["Pending Sales Order Report"]._pending_cache[key];

        if (!cached) {
            frappe.msgprint(__("Could not load pending data for this row."));
            return;
        }

        // cached.rows comes from Python's frappe.as_json(), so it is a
        // JSON string, not an array yet - parse it here before use.
        let rows = cached.rows;
        if (typeof rows === "string") {
            try {
                rows = JSON.parse(rows);
            } catch (e) {
                console.error("Failed to parse pending_popup JSON", e);
                rows = [];
            }
        }

        frappe.query_reports["Pending Sales Order Report"].show_popup(
            rows, cached.customer, cached.sales_order
        );
    },

    show_popup(rows, customer, sales_order) {

        // Escape every dynamic value before inserting into HTML so that
        // apostrophes, quotes, &, <, > in customer/item names never break
        // the popup markup either.
        const esc = (val) => frappe.utils.escape_html(val == null ? "" : String(val));

        let html = `
        <div id="pending-popup">

        <h4>Customer: ${esc(customer)}</h4>
        <h5>Sales Order: ${esc(sales_order)}</h5>

        <div id="print-container">
        <table class="table table-bordered">
        <tr>
            <th>Item Code</th>
            <th>Item Name</th>
            <th>Pending</th>
            <th>Available</th>
            <th>Date</th>
        </tr>`;

        rows.forEach(r => {

            if (r.pending_qty <= 0) return;

            html += `
            <tr>
                <td>${esc(r.item_code)}</td>
                <td>${esc(r.item_name)}</td>
                <td style="color:red;font-weight:bold">${esc(r.pending_qty)}</td>
                <td style="color:green;font-weight:bold">${esc(r.available_qty)}</td>
                <td>${esc(r.delivery_date)}</td>
            </tr>`;
        });

        html += `
        </table>
        </div>

        <button class="btn btn-primary" onclick="print_pending_popup()">Print</button>
        </div>`;

        frappe.msgprint({
            title: "Pending Items",
            message: html,
            wide: true
        });
    }
};


window.print_pending_popup = function () {

    let popup = document.getElementById("pending-popup").cloneNode(true);

    let btn = popup.querySelector("button");
    if (btn) btn.remove();

    let content = popup.outerHTML;

    let w = window.open("", "_blank");

    w.document.open();

    w.document.write(`
        <!DOCTYPE html>
        <html>
        <head>
            <title>Pending Items</title>

            <style>

                @page {
                    size: A4;
                    margin: 10mm;
                }

                body{
                    font-family: Arial, sans-serif;
                    margin:0;
                    width:100%;
                }

                h4{
                    margin-bottom:5px;
                }

                h5{
                    margin-top:0;
                    margin-bottom:15px;
                }

                table{
                    width:100%;
                    border-collapse:collapse;
                    table-layout:auto;
                }

                th, td{
                    border:1px solid black;
                    padding:8px;
                    text-align:left;
                    word-wrap:break-word;
                }

                th{
                    background:#f2f2f2;
                }

            @media print{

                body{
                    width:100%;
                }

                #print-container{
                    max-height: none !important;
                    overflow: visible !important;
                }

                table{
                    width:100%;
                    border-collapse: collapse;
                }

                tr{
                    page-break-inside: avoid;
                }

                thead{
                    display: table-header-group;
                }

            }

            </style>

        </head>

        <body>

        ${content}

        </body>
        </html>
    `);

    w.document.close();

    w.onload = function () {
        w.focus();
        w.print();
    };
};


